File: frontend/src/app/pages/finances/services/credit-card-bill.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';

import { lastValueFrom, take } from 'rxjs';

import dayjs from 'dayjs';

import { CreditCardBillDto } from '../../../models/generated/com/ynixt/sharedfinances/application/web/dto/wallet/creditCard';
import { UserService } from '../../../services/user.service';
import { ONLY_DATE_FORMAT, skipWeekend } from '../../../util/date-util';
import { UserMissingError } from '../errors/user-missing.error';

@Injectable({ providedIn: 'root' })
export class CreditCardBillService {
  private readonly httpClient = inject(HttpClient);
  private readonly userService = inject(UserService);

  getBestBill(
    transactionDate: Date | dayjs.Dayjs,
    dueDay: number,
    dueOnNextBusinessDay: boolean,
    daysBetweenDueAndClosing: number,
  ): dayjs.Dayjs {
    let date = dayjs(transactionDate).startOf('month').date(dueDay);

    if (dueOnNextBusinessDay) {
      date = skipWeekend(date);
    }

    date = date.subtract(daysBetweenDueAndClosing, 'day');

    if (dayjs(transactionDate).isAfter(date)) {
      date = date.startOf('month').add(1, 'month');
    }

    return date.startOf('month');
  }

  async getBillForMonth(creditCardId: string, month: number, year: number): Promise<CreditCardBillDto> {
    const user = await this.userService.getUser();

    if (user != null) {
      return lastValueFrom(
        this.httpClient.get<CreditCardBillDto>(`/api/credit-card-bills/${creditCardId}/of/${year}/${month}`).pipe(take(1)),
      );
    }

    throw new UserMissingError();
  }

  async changeClosingDate(creditCardId: string, closingDate: dayjs.Dayjs): Promise<void> {
    const user = await this.userService.getUser();

    if (user != null) {
      await lastValueFrom(
        this.httpClient
          .put<void>(`/api/credit-card-bills/${creditCardId}/closingDate/${closingDate.format(ONLY_DATE_FORMAT)}`, undefined)
          .pipe(take(1)),
      );
      return;
    }

    throw new UserMissingError();
  }

  async changeDueDate(creditCardId: string, dueDate: dayjs.Dayjs): Promise<void> {
    const user = await this.userService.getUser();

    if (user != null) {
      await lastValueFrom(
        this.httpClient
          .put<void>(`/api/credit-card-bills/${creditCardId}/dueDate/${dueDate.format(ONLY_DATE_FORMAT)}`, undefined)
          .pipe(take(1)),
      );
      return;
    }

    throw new UserMissingError();
  }
}

File: backend/src/main/kotlin/com/ynixt/sharedfinances/domain/services/CreditCardBillService.kt
package com.ynixt.sharedfinances.domain.services

import com.ynixt.sharedfinances.domain.entities.wallet.entries.CreditCardBillEntity
import com.ynixt.sharedfinances.domain.models.creditcard.CreditCardBill
import java.math.BigDecimal
import java.time.LocalDate
import java.util.UUID

interface CreditCardBillService {
    /**
     * Try to get a bill using credit card id and bill date (bill date is dueDate at start of month).
     * If not found, creates a new bill.
     */
    suspend fun getOrCreateBill(
        creditCardId: UUID,
        dueDate: LocalDate,
        closingDate: LocalDate,
        startValue: BigDecimal = BigDecimal.ZERO,
    ): CreditCardBillEntity

    suspend fun changeClosingDate(
        userId: UUID,
        creditCardId: UUID,
        closingDate: LocalDate,
    )

    suspend fun changeDueDate(
        userId: UUID,
        creditCardId: UUID,
        dueDate: LocalDate,
    )

    suspend fun addValueById(
        id: UUID,
        value: BigDecimal,
    ): Long

    suspend fun findById(id: UUID): CreditCardBill?

    suspend fun getBillFromDatabaseOrSimulate(
        userId: UUID,
        creditCardId: UUID,
        billDate: LocalDate,
    ): CreditCardBill
}
